src/NGen.cpp

```diff
@@ -53,12 +68,166 @@ pdm03_struct get_et_params() {
+ *
+ * @param status The initial individual state for the current MPI rank.
+ * @return Whether all ranks coordinating status had a success/ready status value.
+ * @see mpiSyncStatusAnd(bool, const std::string&)
+ */
+bool mpiSyncStatusAnd(bool status) {
+    return mpiSyncStatusAnd(status, "");
+}
+
+/**
+ * Check whether the parameter hydrofabric files have been subdivided into appropriate per partition files.
+ *
+ * Checks to see if partition specific subfiles corresponding to each partition/process already exist.  E.g., for a file
+ * at ``/dirname/catchment_data.geojson`` and two MPI processes, checks if both ``/dirname/catchment_data.geojson.0``
+ * and ``/dirname/catchment_data.geojson.1 `` exist.
+ *
+ * This check is performed for both the catchment and nexus hydrofabric base file names, as stored in the global
+ * ``catchmentDataFile`` and ``nexusDataFile`` variables respectively.  The number of MPI processes is obtained from the
+ * global ``mpi_rank`` variable.
+ *
+ * @param printMessage Whether a supplemental message should be printed to standard out indicating status.
+ *
+ * @return Whether proprocessing has already been performed to divide the main hydrofabric into existing, individual
+ *         sub-hydrofabric files for each partition/process.
+ */
+bool is_hydrofabric_subdivided(bool printMsg) {
+    std::string name = catchmentDataFile + "." + std::to_string(mpi_rank);
+    // Initialize isGood based on local state.  Here, local file is "good" when it already exists.
+    // TODO: this isn't actually checking whether the files are right (just that they are present) so do we need to?
+    bool isGood = utils::FileChecker::file_is_readable(name);
+    if (mpiSyncStatusAnd(isGood)) {
+        if (printMsg) { std::cout << "Hydrofabric already subdivided in " << mpi_num_procs << " files." << std::endl; }
+        return true;
+    }
+    else {
+        if (printMsg) { std::cout << "Hydrofabric has not yet been subdivided." << std::endl; }
+        return false;
+    }
+}
+
+/**
+ * Convenience overloaded method for when no supplemental output message is required.
+ *
+ * @return Whether proprocessing has already been performed to divide the main hydrofabric into existing, individual
+ *         sub-hydrofabric files for each partition/process.
+ * @see is_hydrofabric_subdivided(bool)
+ */
+bool is_hydrofabric_subdivided() {
+    return is_hydrofabric_subdivided(false);
+}
+
+/**
+ * Attempt to subdivide the passed hydrofabric files into a series of per-partition files.
+ *
+ * This function assumes that, when it is called, the intent is for it to produce a freshly subdivided hydrofabric and
+ * associated files.  As a result, if there are any other subdivided hydrofabric files present having the same names
+ * as the files the function will write, then those preexisting files are considered stale and overwritten.
+ *
+ * @return Whether subdividing was successful.
+ */
+bool subdivide_hydrofabric() {
```

Comment:
Wondering if a `parallel_utils` would be a good place for a few of these function definitions.

Reply:
Moved to dedicated file.